### PR TITLE
[SDAG][ISel][TableGen][LoongArch] Report error for trivial bitcasts when there are predicate calls

### LINKED ISSUE #116008: [LoongArch] Miscompilation
Reproducer: https://godbolt.org/z/G4c4Yvcrx
```
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "loongarch64-unknown-linux-gnu"

@f = dso_local local_unnamed_addr global i32 5, align 4

; Function Attrs: nofree nounwind
define dso_local noundef signext i32 @main() {
entry:
  %1 = load i32, ptr @f, align 4
  %.fr = freeze i32 %1
  %10 = insertelement <4 x i32> poison, i32 %.fr, i64 0
  %11 = shufflevector <4 x i32> %10, <4 x i32> poison, <4 x i32> zeroinitializer
  %12 = icmp ugt <4 x i32> zeroinitializer, %11
  %13 = select <4 x i1> %12, <4 x i32> splat (i32 1), <4 x i32> %11
  %14 = icmp samesign ugt <4 x i32> %13, splat (i32 31)
  %15 = lshr <4 x i32> splat (i32 2147483647), %13
  %16 = icmp samesign ugt <4 x i32> %11, %15
  %17 = select <4 x i1> %14, <4 x i1> splat (i1 true), <4 x i1> %16
  %18 = select <4 x i1> %17, <4 x i32> zeroinitializer, <4 x i32> %13
  %19 = shl <4 x i32> %11, %18
  %20 = xor <4 x i32> %19, splat (i32 1)
  %21 = tail call i32 @llvm.vector.reduce.or.v4i32(<4 x i32> %20)
  ret i32 %21
}
```
```
> bin/lli test.ll
> echo $?
33
```
But the correct exit code should be 161 (run with lli on x86/https://github.com/dtcxzyw/llvm-ub-aware-interpreter):
```
; llubi test.ll --verbose
Entering function main
  %0 = load i32, ptr @f, align 4 -> i32 5
  %.fr = freeze i32 %0 -> i32 5
  %1 = insertelement <4 x i32> poison, i32 %.fr, i64 0 -> { i32 5, poison, poison, poison }
  %2 = shufflevector <4 x i32> %1, <4 x i32> poison, <4 x i32> zeroinitializer -> { i32 5, i32 5, i32 5, i32 5 }
  %3 = icmp ugt <4 x i32> zeroinitializer, %2 -> { F, F, F, F }
  %4 = select <4 x i1> %3, <4 x i32> splat (i32 1), <4 x i32> %2 -> { i32 5, i32 5, i32 5, i32 5 }
  %5 = icmp samesign ugt <4 x i32> %4, splat (i32 31) -> { F, F, F, F }
  %6 = lshr <4 x i32> splat (i32 2147483647), %4 -> { i32 67108863, i32 67108863, i32 67108863, i32 67108863 }
  %7 = icmp samesign ugt <4 x i32> %2, %6 -> { F, F, F, F }
  %8 = select <4 x i1> %5, <4 x i1> splat (i1 true), <4 x i1> %7 -> { F, F, F, F }
  %9 = select <4 x i1> %8, <4 x i32> zeroinitializer, <4 x i32> %4 -> { i32 5, i32 5, i32 5, i32 5 }
  %10 = shl <4 x i32> %2, %9 -> { i32 160, i32 160, i32 160, i32 160 }
  %11 = xor <4 x i32> %10, splat (i32 1) -> { i32 161, i32 161, i32 161, i32 161 }
  %12 = tail call i32 @llvm.vector.reduce.or.v4i32(<4 x i32> %11) -> i32 161
  ret i32 %12
Exiting function main
```
llvm version: edfa75de33433de29f438fbea4145ec6ae20e020
CPU: Loongson-3A5000

I will post a fix later.

### PR DESCRIPTION
**Pull request by dtcxzyw**
On loongarch64 with lsx extension, we select `VBITREV_W` for  `v4i32 (xor X, (shl splat(1), Y))`:
https://github.com/llvm/llvm-project/blob/8e6630391699116641cf390a10476295b7d4b95c/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td#L1583-L1584

And `vsplat_imm_eq_1` is defined as:
https://github.com/llvm/llvm-project/blob/8e6630391699116641cf390a10476295b7d4b95c/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td#L77-L87

For the `(bitconvert (v4i32 (build_vector)))` case, the pattern is expected to be:
```
PATTERN: (xor:{ *:[v4i32] } v4i32:{ *:[v4i32] }:$vj, (shl:{ *:[v4i32] } (bitconvert:{ *:[v4i32] } (build_vector:{ *:[v4i32] }))<<P:Predicate_vsplat_imm_eq_1>>, v4i32:{ *:[v4i32] }:$vk))
RESULT:  (VBITREV_W:{ *:[v4i32] } v4i32:{ *:[v4i32] }:$vj, v4i32:{ *:[v4i32] }:$vk)
```

However, `simplifyTree` drops the `bitconvert` node and its predicates:
https://github.com/llvm/llvm-project/blob/8e6630391699116641cf390a10476295b7d4b95c/llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp#L3036-L3062

Then llvm will match `vsplat_imm_eq_1` for any v4i32 splats and cause a miscompilation:
```
PATTERN: (xor:{ *:[v4i32] } v4i32:{ *:[v4i32] }:$vj, (shl:{ *:[v4i32] } (build_vector:{ *:[v4i32] }), v4i32:{ *:[v4i32] }:$vk))
RESULT:  (VBITREV_W:{ *:[v4i32] } v4i32:{ *:[v4i32] }:$vj, v4i32:{ *:[v4i32] }:$vk)
```

This patch adds additional checks for predicates associated with the trivial bitconvert node. Unused patterns in the LoongArch target are also removed.

Fixes https://github.com/llvm/llvm-project/issues/116008.

### CODE REVIEW

**Comment by topperc** [comment]
The (v4i32 (bitconvert (v4i32)) could never exist in the DAG. Is the issue here that we have both a pattern with the bitconvert and a pattern without it due to `(build_vector)` also being listed? Tablegen removing the bitconvert gives 2 duplicate patterns except one no longer has a predicate and gets higher priority?

So I think this patch prevents the bitconvert pattern from being simplified away. Leaving an unneeded pattern that can never match?

**Comment by topperc** [comment]
No tests fail if you just delete the bitconvert from the pattern

```
diff --git a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
index 250896cbbe5f..2caf7e0d1908 100644
--- a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
+++ b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
@@ -75,7 +75,7 @@ class VecCond<SDPatternOperator OpNode, ValueType TyNode,
 }
 
 def vsplat_imm_eq_1 : PatFrags<(ops), [(build_vector),
-                                       (bitconvert (v4i32 (build_vector)))], [{
+                                       ], [{
   APInt Imm;
   EVT EltTy = N->getValueType(0).getVectorElementType();
 

```

**Comment by wangleiat** [comment]
> No tests fail if you just delete the bitconvert from the pattern
> 
> ```
> diff --git a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> index 250896cbbe5f..2caf7e0d1908 100644
> --- a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> +++ b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> @@ -75,7 +75,7 @@ class VecCond<SDPatternOperator OpNode, ValueType TyNode,
>  }
>  
>  def vsplat_imm_eq_1 : PatFrags<(ops), [(build_vector),
> -                                       (bitconvert (v4i32 (build_vector)))], [{
> +                                       ], [{
>    APInt Imm;
>    EVT EltTy = N->getValueType(0).getVectorElementType();
>  
> ```

I agree, I must have been foolish at that time.

**Comment by dtcxzyw** [comment]
> No tests fail if you just delete the bitconvert from the pattern
> 
> ```
> diff --git a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> index 250896cbbe5f..2caf7e0d1908 100644
> --- a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> +++ b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> @@ -75,7 +75,7 @@ class VecCond<SDPatternOperator OpNode, ValueType TyNode,
>  }
>  
>  def vsplat_imm_eq_1 : PatFrags<(ops), [(build_vector),
> -                                       (bitconvert (v4i32 (build_vector)))], [{
> +                                       ], [{
>    APInt Imm;
>    EVT EltTy = N->getValueType(0).getVectorElementType();
>  
> ```

Done.

**Comment by topperc** [comment]
> > No tests fail if you just delete the bitconvert from the pattern
> 
> > 
> 
> > ```
> 
> > diff --git a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> 
> > index 250896cbbe5f..2caf7e0d1908 100644
> 
> > --- a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> 
> > +++ b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> 
> > @@ -75,7 +75,7 @@ class VecCond<SDPatternOperator OpNode, ValueType TyNode,
> 
> >  }
> 
> >  
> 
> >  def vsplat_imm_eq_1 : PatFrags<(ops), [(build_vector),
> 
> > -                                       (bitconvert (v4i32 (build_vector)))], [{
> 
> > +                                       ], [{
> 
> >    APInt Imm;
> 
> >    EVT EltTy = N->getValueType(0).getVectorElementType();
> 
> >  
> 
> > ```
> 
> 
> 
> I agree, I must have been foolish at that time.

The pattern was added before build_vector lowering was added. Maybe it used to be needed?

**Comment by dtcxzyw** [comment]
> The pattern was added before build_vector lowering was added. Maybe it used to be needed?

No tests in https://reviews.llvm.org/D155829. I am not sure if it is safe to remove these patterns.
cc @wangleiat @SixWeining @xen0n

**Comment by dtcxzyw** [comment]
> > > No tests fail if you just delete the bitconvert from the pattern
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > diff --git a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> > 
> > 
> > > index 250896cbbe5f..2caf7e0d1908 100644
> > 
> > 
> > > --- a/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> > 
> > 
> > > +++ b/llvm/lib/Target/LoongArch/LoongArchLSXInstrInfo.td
> > 
> > 
> > > @@ -75,7 +75,7 @@ class VecCond<SDPatternOperator OpNode, ValueType TyNode,
> > 
> > 
> > > }
> > 
> > 
> > > 
> > 
> > 
> > > def vsplat_imm_eq_1 : PatFrags<(ops), [(build_vector),
> > 
> > 
> > > * ```
> > >                                     (bitconvert (v4i32 (build_vector)))], [{
> > >   ```
> > 
> > 
> > > * ```
> > >                                     ], [{
> > >   ```
> > 
> > 
> > > APInt Imm;
> > 
> > 
> > > EVT EltTy = N->getValueType(0).getVectorElementType();
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > I agree, I must have been foolish at that time.
> 
> The pattern was added before build_vector lowering was added. Maybe it used to be needed?

```
[5653/12011] Building MipsGenDAGISel.inc...
FAILED: lib/Target/Mips/MipsGenDAGISel.inc /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenDAGISel.inc
cd /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build && /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/bin/llvm-tblgen -gen-dag-isel -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/include -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/include -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target -omit-comments /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/Mips.td --write-if-changed -o lib/Target/Mips/MipsGenDAGISel.inc -d lib/Target/Mips/MipsGenDAGISel.inc.d
error: (bitconvert:{ *:[v4i32] } (build_vector:{ *:[v4i32] }))<<P:Predicate_vsplati64_imm_eq_1>>
 trivial bitconvert node should not have predicate calls
[5654/12011] Building MipsGenInstrInfo.inc...
FAILED: lib/Target/Mips/MipsGenInstrInfo.inc /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenInstrInfo.inc
cd /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build && /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/bin/llvm-tblgen -gen-instr-info -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/include -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/include -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/Mips.td --write-if-changed -o lib/Target/Mips/MipsGenInstrInfo.inc -d lib/Target/Mips/MipsGenInstrInfo.inc.d
error: (bitconvert:{ *:[v4i32] } (build_vector:{ *:[v4i32] }))<<P:Predicate_vsplati64_imm_eq_1>>
 trivial bitconvert node should not have predicate calls
[5655/12011] Building MipsGenFastISel.inc...
FAILED: lib/Target/Mips/MipsGenFastISel.inc /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenFastISel.inc
cd /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build && /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/bin/llvm-tblgen -gen-fast-isel -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/include -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/include -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/Mips.td --write-if-changed -o lib/Target/Mips/MipsGenFastISel.inc -d lib/Target/Mips/MipsGenFastISel.inc.d
error: (bitconvert:{ *:[v4i32] } (build_vector:{ *:[v4i32] }))<<P:Predicate_vsplati64_imm_eq_1>>
 trivial bitconvert node should not have predicate calls
[5656/12011] Building MipsGenGlobalISel.inc...
FAILED: lib/Target/Mips/MipsGenGlobalISel.inc /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenGlobalISel.inc
cd /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build && /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/bin/llvm-tblgen -gen-global-isel -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/build/include -I/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/include -I /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-fkcjf-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/Mips.td --write-if-changed -o lib/Target/Mips/MipsGenGlobalISel.inc -d lib/Target/Mips/MipsGenGlobalISel.inc.d
error: (bitconvert:{ *:[v4i32] } (build_vector:{ *:[v4i32] }))<<P:Predicate_vsplati64_imm_eq_1>>
 trivial bitconvert node should not have predicate calls
```

I guess these patterns in LoongArch backend are copied from the MIPS backend.

**Comment by arsenm** [comment]
> No tests in https://reviews.llvm.org/D155829. I am not sure if it is safe to remove these patterns.

If there are no tests just remove it

**Comment by topperc** [comment]
The Mips and LoongArch PatFrags would probably be better written as ComplexPatterns.

File: llvm/lib/Target/Mips/MipsMSAInstrInfo.td [hunk 1]
```diff
@@ -197,6 +197,17 @@ def vsplati16 : PatFrag<(ops node:$e0),
 def vsplati32 : PatFrag<(ops node:$e0),
                         (v4i32 (build_vector node:$e0, node:$e0,
                                              node:$e0, node:$e0))>;
+// Any build_vector that is a constant splat with a value that equals 1
+// FIXME: These should be a ComplexPattern but we can't use them because the
```
Comment (topperc) [inline llvm/lib/Target/Mips/MipsMSAInstrInfo.td:201]: I'm going to try to resolve this FIXME.

Comment (topperc) [reply llvm/lib/Target/Mips/MipsMSAInstrInfo.td:201]: #116471

Comment (dtcxzyw) [reply llvm/lib/Target/Mips/MipsMSAInstrInfo.td]: Rebased.

**Comment by dtcxzyw** [comment]
Ping @wangleiat

**Review by wangleiat — APPROVED** [review]
LGTM, thanks! Could you share how you discovered this issue? Also, can this patch be cherry-picked to 19?

**Comment by dtcxzyw** [comment]
> Could you share how you discovered this issue? 

I discovered this issue with [a modified version of csmith](https://github.com/dtcxzyw/csmith) running on a Loongson-3A5000-based machine.

> Also, can this patch be cherry-picked to 19?

Sure.

**Comment by dtcxzyw** [comment]
@topperc Any further comments?

**Comment by wangleiat** [comment]
> > Could you share how you discovered this issue?
> 
> I discovered this issue with [a modified version of csmith](https://github.com/dtcxzyw/csmith) running on a Loongson-3A5000-based machine.
> 
> > Also, can this patch be cherry-picked to 19?
> 
> Sure.

Thanks!

**Comment by topperc** [comment]
> > Could you share how you discovered this issue?
> 
> I discovered this issue with [a modified version of csmith](https://github.com/dtcxzyw/csmith) running on a Loongson-3A5000-based machine.
> 
> > Also, can this patch be cherry-picked to 19?
> 
> Sure.

We can't cherry-pick the tablegen code without my Mips changes though right?

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
> > > Could you share how you discovered this issue?
> > 
> > 
> > I discovered this issue with [a modified version of csmith](https://github.com/dtcxzyw/csmith) running on a Loongson-3A5000-based machine.
> > > Also, can this patch be cherry-picked to 19?
> > 
> > 
> > Sure.
> 
> We can't cherry-pick the tablegen code without my Mips changes though right?

I will cherry-pick https://github.com/llvm/llvm-project/commit/2f4572f5e7e2d7f4626e825404c11f07d191fb05 as well.

**Comment by dtcxzyw** [comment]
/cherry-pick 2f4572f5e7e2d7f4626e825404c11f07d191fb05 c727b48287cc96888f9e262f23d53cf635cf3b3d

**Pull request merged**